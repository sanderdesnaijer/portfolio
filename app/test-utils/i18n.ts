import { AbstractIntlMessages } from "next-intl";
import translations from "../translations/en.json";

export type MockTranslations = typeof translations;

export const getTranslationKey = (key: keyof MockTranslations): string => {
  return ((translations as AbstractIntlMessages)[key] || key) as string;
};
