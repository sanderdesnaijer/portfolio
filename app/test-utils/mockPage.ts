import { PageSanity } from "@/sanity/types";

export const mockPage: PageSanity = {
  title: "My page",
  imageAlt: "Profile Picture",
  imageURL: "mock-image-url",
  body: [
    {
      _type: "block",
      children: [
        {
          text: "Mock body content",
          _type: "span",
          marks: [],
        },
      ],
      style: "",
    },
  ],
  name: "",
  description: "",
  slug: {
    _type: "slug",
    current: "page-slug",
  },
  mainImage: {
    _type: "image",
    asset: {
      _ref: "",
      _type: "",
    },
    alt: undefined,
  },
  publishedAt: "",
  _id: "page-1",
  _rev: "",
  _type: "",
  _createdAt: "2025-01-01",
  _updatedAt: "2025-01-02",
};

const mockPage2: PageSanity = {
  title: "My page 2",
  imageAlt: "Page 2 Picture",
  imageURL: "mock-image-url",
  body: [
    {
      _type: "block",
      children: [
        {
          text: "Mock body content page 2",
          _type: "span",
          marks: [],
        },
      ],
      style: "",
    },
  ],
  name: "",
  description: "",
  slug: {
    _type: "slug",
    current: "page2-slug",
  },
  mainImage: {
    _type: "image",
    asset: {
      _ref: "",
      _type: "",
    },
    alt: undefined,
  },
  publishedAt: "",
  _id: "page-2",
  _rev: "",
  _type: "",
  _createdAt: "",
  _updatedAt: "",
};

export const mockPages: PageSanity[] = [mockPage, mockPage2];
