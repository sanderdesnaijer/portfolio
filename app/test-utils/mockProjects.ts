import { ProjectTypeSanity } from "@/sanity/types";

export const mockProject: ProjectTypeSanity = {
  _id: "1",
  title: "Project 1",
  _createdAt: "2025-01-01",
  slug: { _type: "slug", current: "/project-1" },
  imageURL: "http://mocked-image-url.com/image1.jpg",
  mainImage: {
    _type: "image",
    alt: "Project 1 Image Alt",
    asset: { _ref: "string", _type: "ref" },
  },
  body: [
    {
      _type: "block",
      children: [{ text: "Mock body content", _type: "span", marks: [] }],
      style: "",
    },
  ],
  links: [
    { title: "GitHub", icon: "github", link: "http://test" },
    { title: "Live Demo", icon: "external-link", link: "http://test" },
  ],
  publishedAt: "",
  _updatedAt: "",
  _rev: "",
  _type: "",
};

export const mockProjects: ProjectTypeSanity[] = [
  mockProject,
  {
    _id: "2",
    title: "Project 2",
    _createdAt: "2025-01-02",
    slug: { current: "/project-2", _type: "slug" },
    imageURL: "http://mocked-image-url.com/image2.jpg",
    mainImage: {
      _type: "image",
      alt: "Project 2 Image Alt",
      asset: { _ref: "string", _type: "ref" },
    },
    body: null!,
    links: [],
    publishedAt: "",
    _updatedAt: "",
    _rev: "",
    _type: "",
  },
];
