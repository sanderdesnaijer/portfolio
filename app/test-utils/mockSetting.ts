import { SettingSanity } from "@/sanity/types";

export const mockSetting: SettingSanity = {
  _id: "3e37e065-e582-450f-bba0-70709082f578",
  title: "Sander de Snaijer",
  description:
    "Passionate software developer turning creative ideas to functional products.",
  socialMedia: [
    {
      title: "Github",
      link: "https://github.com/sanderdesnaijer",
      icon: "github",
    },
    {
      title: "Linkedin",
      link: "https://www.linkedin.com/in/sanderdesnaijer/",
      icon: "linkedin",
    },
    {
      title: "Gitlab",
      link: "https://gitlab.com/sanderdesnaijer",
      icon: "gitlab",
    },
  ],
  _rev: "",
  _type: "",
  _createdAt: "",
  _updatedAt: "",
  imageURL: "",
  imageAlt: "",
};
