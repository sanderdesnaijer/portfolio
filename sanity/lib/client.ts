import { createClient } from "next-sanity";

import { apiVersion, dataset, projectId } from "../env";
import envConfig from "@/envConfig";

export const client = createClient({
  projectId,
  dataset,
  apiVersion,
  token: envConfig.sanityToken,
  useCdn: process.env.NODE_ENV === "production", // Use CDN in production
});
