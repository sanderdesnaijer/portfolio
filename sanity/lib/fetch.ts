import "server-only";

import type { QueryParams } from "@sanity/client";
import { draftMode } from "next/headers";
import { client } from "@/sanity/lib/client";
import { REVALIDATE_INTERVAL } from "@/app/utils/constants";

const DEFAULT_PARAMS = {} as QueryParams;
const DEFAULT_TAGS = [] as string[];

export const token = process.env.SANITY_API_READ_TOKEN;

export async function sanityFetch<QueryResponse>({
  query,
  params = DEFAULT_PARAMS,
  tags = DEFAULT_TAGS,
}: {
  query: string;
  params?: QueryParams;
  tags?: string[];
}): Promise<QueryResponse> {
  const currentDraftMode = await draftMode();
  const isDraftMode = currentDraftMode.isEnabled;
  const queryParams = await params;

  if (isDraftMode && !token) {
    throw new Error(
      "The `SANITY_API_READ_TOKEN` environment variable is required."
    );
  }
  const isDevelopment = process.env.NODE_ENV === "development";

  return client
    .withConfig({
      useCdn: isDevelopment,
    })
    .fetch<QueryResponse>(query, queryParams, {
      cache: isDevelopment || isDraftMode ? undefined : "force-cache",
      // cache: isDevelopment || isDraftMode ? undefined : "no-store",
      ...(isDraftMode && {
        token,
        perspective: "previewDrafts",
      }),
      next: {
        revalidate:
          isDevelopment || isDraftMode ? undefined : REVALIDATE_INTERVAL,
        tags,
      },
    });
}
