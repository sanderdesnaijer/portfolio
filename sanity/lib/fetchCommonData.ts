import { sanityFetch } from "@/sanity/lib/fetch";
import { allPagesQuery, settingsQuery } from "@/sanity/lib/queries";
import { PageSanity, SettingSanity } from "@/sanity/types";
import { MenuItem } from "@/app/components/Menu";

export async function fetchCommonData() {
  const setting = await sanityFetch<SettingSanity>({ query: settingsQuery });
  const pages = await sanityFetch<PageSanity[]>({ query: allPagesQuery })
    .then((res) => {
      return res || [];
    })
    .catch(() => {
      // console.log(error);
      return [];
    });
  const menuItems = pages.map<MenuItem>(({ title, slug }) => ({
    title,
    pathname: `/${slug && slug.current ? slug.current : ""}`,
  }));

  return { setting, menuItems };
}
