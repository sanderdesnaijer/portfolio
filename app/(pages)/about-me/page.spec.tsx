import { render, screen } from "@testing-library/react";
import { sanityFetch } from "@/sanity/lib/fetch";
import Page from "./page";
import { getTranslationKey } from "@/app/test-utils/i18n";

jest.mock("next-sanity", () => ({
  ...jest.requireActual("next-sanity"),
  groq: (query: string) => query,

  PortableText: ({ value }: { value: unknown }) => (
    <div data-testid="portable-text">{JSON.stringify(value)}</div>
  ),
}));

describe("Page Component", () => {
  const mockPage = {
    title: "About Me",
    imageAlt: "Profile Picture",
    imageURL: "mock-image-url",
    body: [{ type: "text", children: [{ text: "Mock body content" }] }],
  };

  const mockJobs = [
    {
      _id: "job-1",
      imageURL: "mock-job-image-url",
      companyName: "Company A",
      jobTitle: "Developer",
      startDate: "2022-01-01",
      endDate: null,
      description: [{ type: "text", children: [{ text: "Job description" }] }],
      links: [
        {
          title: "GitHub",
          icon: "github",
        },
      ],
      tags: "React, TypeScript",
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders the page with correct data", async () => {
    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockPage) // First call for page data
      .mockResolvedValueOnce(mockJobs); // Second call for jobs data

    const { container } = render(await Page());

    // Check page title
    expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
      "About Me"
    );

    // Check image rendering
    expect(
      screen.getByAltText("Profile Picture").getAttribute("src")
    ).toContain("mocked-image-url");

    // Check body content rendering
    expect(screen.getByText(JSON.stringify(mockPage.body))).toBeTruthy();

    // Check job experience section
    expect(screen.getByText("Company A")).toBeInTheDocument();
    expect(screen.getByText("Developer")).toBeInTheDocument();
    expect(
      screen.getByText(`Jan 2022 - ${getTranslationKey("date-present")}`)
    ).toBeInTheDocument();

    // Check links in job
    expect(screen.getByText("GitHub")).toBeInTheDocument();

    // Check snapshot
    expect(container).toMatchSnapshot();
  });

  it("renders [Page not found] when page data is missing", async () => {
    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(null) // First call for page data
      .mockResolvedValueOnce(mockJobs); // Second call for jobs data

    render(await Page());

    expect(screen.getByText("[Page not found]")).toBeInTheDocument();
  });

  it("handles missing job links gracefully", async () => {
    const mockJobsWithoutLinks = [
      {
        ...mockJobs[0],
        links: null, // Remove links
      },
    ];

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockPage) // First call for page data
      .mockResolvedValueOnce(mockJobsWithoutLinks); // Second call for jobs data

    render(await Page());

    // Ensure it doesn't throw an error and renders correctly
    expect(screen.queryByText("GitHub")).not.toBeInTheDocument();
  });

  it("handles empty jobs list gracefully", async () => {
    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockPage) // First call for page data
      .mockResolvedValueOnce([]); // Second call for jobs data

    render(await Page());

    // Ensure no jobs are rendered
    expect(screen.queryByRole("listitem")).not.toBeInTheDocument();
  });

  it("renders correctly when page body is missing", async () => {
    const mockPageWithoutBody = {
      ...mockPage,
      body: null, // Body is missing
    };

    (sanityFetch as jest.Mock).mockResolvedValueOnce(mockPageWithoutBody); // First call for page data

    render(await Page());

    // Ensure the PortableText is not rendered
    expect(screen.queryByTestId("portable-text")).not.toBeInTheDocument();

    // Verify the page title still renders
    expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
      "About Me"
    );
  });

  it("renders job with endDate correctly", async () => {
    const mockJobsWithEndDate = [
      {
        ...mockJobs[0],
        endDate: "2023-12-31", // Provide an endDate
      },
    ];

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockPage) // First call for page data
      .mockResolvedValueOnce(mockJobsWithEndDate); // Second call for jobs data

    render(await Page());

    // Check that the job's start and end date is rendered correctly
    expect(screen.getByText("Jan 2022 - Dec 2023")).toBeInTheDocument();
  });

  it("renders [present] when job endDate is not given", async () => {
    const mockJobsWithoutEndDate = [
      {
        ...mockJobs[0],
        endDate: null, // No end date
      },
    ];

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockPage) // First call for page data
      .mockResolvedValueOnce(mockJobsWithoutEndDate); // Second call for jobs data

    render(await Page());

    // Check that "[present]" is displayed for the job
    expect(
      screen.getByText(`Jan 2022 - ${getTranslationKey("date-present")}`)
    ).toBeInTheDocument();
  });
});
