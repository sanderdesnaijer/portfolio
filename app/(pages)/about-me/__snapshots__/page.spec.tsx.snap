// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Component renders the page with correct data 1`] = `
<div>
  <main
    class="container mx-auto prose prose-xl px-4 py-16 dark:prose-invert"
  >
    <h1>
      About Me
    </h1>
    <img
      alt="Profile Picture"
      data-nimg="1"
      decoding="async"
      height="300"
      src="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=640&q=75"
      srcset="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=384&q=75 1x, /_next/image?url=http%3A%2F%2Fmocked-image-url&w=640&q=75 2x"
      style="color: transparent;"
      width="300"
    />
    <div
      data-testid="portable-text"
    >
      [{"type":"text","children":[{"text":"Mock body content"}]}]
    </div>
    <h3>
      Experience
    </h3>
    <ol>
      <li>
        <img
          alt="mock-job-image-url"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=128&q=75"
          srcset="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=64&q=75 1x, /_next/image?url=http%3A%2F%2Fmocked-image-url&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
        <h3>
          Company A
        </h3>
        <p>
          Developer
        </p>
        <p
          class=""
        >
          Jan 2022 - present
        </p>
        <div
          data-testid="portable-text"
        >
          [{"type":"text","children":[{"text":"Job description"}]}]
        </div>
        <ul>
          <li
            aria-label="github icon"
            title="github"
          >
            svgrurl
            <h3>
              GitHub
            </h3>
          </li>
        </ul>
        <p>
          React, TypeScript
        </p>
      </li>
    </ol>
  </main>
</div>
`;

exports[`app/(pages)/about-me renders the page with correct data 1`] = `
<div>
  <main
    class="container mx-auto prose prose-xl px-4 py-16 dark:prose-invert"
  >
    <h1>
      About Me
    </h1>
    <img
      alt="Profile Picture"
      data-nimg="1"
      decoding="async"
      height="300"
      src="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=640&q=75"
      srcset="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=384&q=75 1x, /_next/image?url=http%3A%2F%2Fmocked-image-url&w=640&q=75 2x"
      style="color: transparent;"
      width="300"
    />
    <div
      data-testid="portable-text"
    >
      [{"_type":"block","children":[{"text":"Mock body content","_type":"span","marks":[]}],"style":""}]
    </div>
    <h3>
      Experience
    </h3>
    <ol>
      <li>
        <img
          alt="mock-job-image-url"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=128&q=75"
          srcset="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=64&q=75 1x, /_next/image?url=http%3A%2F%2Fmocked-image-url&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
        <h3>
          Company A
        </h3>
        <p>
          Developer
        </p>
        <p
          class=""
        >
          Jan 2022 - present
        </p>
        <div
          data-testid="portable-text"
        >
          [{"_type":"block","children":[{"text":"Job description","_type":"span","marks":[]}],"style":""}]
        </div>
        <ul>
          <li
            aria-label="github icon"
            title="github"
          >
            svgrurl
            <h3>
              GitHub
            </h3>
          </li>
        </ul>
        <p>
          React, TypeScript
        </p>
      </li>
    </ol>
  </main>
</div>
`;
